File: Python/BOJ/시뮬레이션/16236.py
//
//  16236.py
//  Algorithm
//
//  Created by 유정주 on 2023/06/07.
//

import sys; readLine = sys.stdin.readline

dx = [-1, 1, 0, 0]; dy = [0, 0, -1, 1]

n = int(readLine())
board = [list(map(int, readLine().split())) for _ in range(n)]

sharkSize = 2
for i in range(n):
  for j in range(n):
    if board[i][j] == 9:
      shark = (i, j)
      board[i][j] = 0
      
result = 0
eatCount = 0
while True:
  canEat = False

  q = [(shark[0], shark[1], 0)]
  index = 0
  visited = [[False for _ in range(n)] for _ in range(n)]
  visited[shark[0]][shark[1]] = True

  eatTime = 0
  while index < len(q):
    x, y, time = q[index]
    index += 1

    item = board[x][y]
    if item > 0 and item < sharkSize:
      canEat = True
      eatTime = time
      break
    
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if not (0 <= nx < n and 0 <= ny < n and not visited[nx][ny]):
        continue

      visited[nx][ny] = True
      if board[nx][ny] <= sharkSize:
        q.append((nx, ny, time + 1))
  
  if canEat:
    fishes = []
    for x, y, time in q:
      if time == eatTime and board[x][y] > 0 and board[x][y] < sharkSize:
        fishes.append((x, y))
    fishes.sort(key = lambda x: (x[0], x[1]))
    
    eattedFish = fishes[0]
    board[eattedFish[0]][eattedFish[1]] = 0
    shark = (eattedFish[0], eattedFish[1])

    result += eatTime
    eatCount += 1
    if sharkSize == eatCount:
      sharkSize += 1
      eatCount = 0
  else:
    break

print(result)

File: Swift/BOJ/16000/16236.swift
//
//  16236.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/08/16.
//

import Foundation

typealias Point = (x: Int, y: Int, time: Int)

let dx = [-1, 1, 0, 0]
let dy = [0, 0, -1, 1]


let n = Int(readLine()!)!
var map: [[Int]] = []
for _ in 0..<n {
    map.append(readLine()!.split { $0 == " " }.map { Int(String($0))! })
}

var shark: Point = (0, 0, 0)
var size = 2, eatCount = 0
var result = 0

for i in 0..<n {
    for j in 0..<n {
        if map[i][j] == 9 {
            shark.x = i
            shark.y = j
        }
    }
}

while true {
    var queue: [Point] = [shark]
    var index = 0
    var visited: [[Bool]] = Array(repeating: Array(repeating: false, count: n), count: n)
    visited[shark.x][shark.y] = true
    
    var fishes: [Point] = []
    
    while index < queue.count {
        let item = queue[index]
        index += 1
        
        if let first = fishes.first, item.time >= first.time {
            break
        }
        
        for i in 0..<4 {
            let nx = item.x + dx[i]
            let ny = item.y + dy[i]
            let nt = item.time + 1
            
            guard 0..<n ~= nx, 0..<n ~= ny, !visited[nx][ny], map[nx][ny] <= size else {
                continue
            }
            visited[nx][ny] = true
            
            if map[nx][ny] > 0 && map[nx][ny] < size {
                fishes.append((nx, ny, nt))
            }
            
            queue.append((nx, ny, nt))
        }
    }
    
    if fishes.isEmpty {
        break
    }
    
    fishes.sort {
        $0.time != $1.time ?
            $0.time < $1.time :
            ($0.x != $1.x ? $0.x < $1.x : $0.y < $1.y)
    }
    
    let eattedFish = fishes.first!
    map[shark.x][shark.y] = 0
    
    shark.x = eattedFish.x
    shark.y = eattedFish.y
    shark.time = 0
    
    result += eattedFish.time
    eatCount += 1
    if eatCount == size {
        size += 1
        eatCount = 0
    }
}

print(result)
